orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/artifacts/FindArtifactsFromResourceTask.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2018 Joel Wilsson
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.netflix.spinnaker.orca.clouddriver.tasks.artifacts;
+
+import com.netflix.spinnaker.kork.artifacts.model.Artifact;
+import com.netflix.spinnaker.kork.artifacts.model.ExpectedArtifact;
+import com.netflix.spinnaker.orca.ExecutionStatus;
+import com.netflix.spinnaker.orca.Task;
+import com.netflix.spinnaker.orca.TaskResult;
+import com.netflix.spinnaker.orca.clouddriver.OortService;
+import com.netflix.spinnaker.orca.clouddriver.model.Manifest;
+import com.netflix.spinnaker.orca.clouddriver.tasks.AbstractCloudProviderAwareTask;
+import com.netflix.spinnaker.orca.pipeline.model.Stage;
+import com.netflix.spinnaker.orca.pipeline.util.ArtifactResolver;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import javax.annotation.Nonnull;
+import java.util.*;
+import java.util.stream.Collectors;
+
+@Slf4j
+@Component
+public class FindArtifactsFromResourceTask extends AbstractCloudProviderAwareTask implements Task {
+  public static final String TASK_NAME = "findArtifactsFromResource";
+
+  @Autowired
+  ArtifactResolver artifactResolver;
+
+  @Autowired
+  OortService oortService;
+
+  @Nonnull
+  @Override
+  public TaskResult execute(@Nonnull Stage stage) {
+    StageData stageData = stage.mapTo(StageData.class);
+    String account = getCredentials(stage);
+    Set<String> expectedArtifactIds = new HashSet<>(stageData.expectedArtifactIds);
+    List<ExpectedArtifact> expectedArtifacts = artifactResolver.getExpectedArtifacts(stage);
```

Comment:
I don't think you need to be looking up every expected artifact here - from my reading you're looking all the way back into the stage history to find every expected artifact declared in upstream stages.

I would think that all you need are the expected artifacts declared in this stage.

Reply:
Maybe, I'm not sure exactly when the binding happen, but I need to get a `List<ExpectedArtifact>` to check if they are bound, and if so, if they are bound to the default artifact. This was the only way I could find.

https://github.com/spinnaker/orca/pull/1914/files#diff-ab3dd9a63d59c02eb0cf7191ce664c72R54 gets me a `List<ExpectedArtifact>` but it takes them from `stage.getContext()).getAll("resolvedExpectedArtifacts")`, while I also want unresolved ones (as in the Docker image from Configuration without a default).